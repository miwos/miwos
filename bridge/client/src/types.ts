export type MessageArg = string | number | boolean | Uint8Array

export interface Message {
  address: string
  args: MessageArg[]
}

export interface DirItem {
  name: string
  type: 'directory' | 'file'
  children?: DirItem[]
}

export type Dir = DirItem[]
