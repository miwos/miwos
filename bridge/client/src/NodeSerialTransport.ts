import type { Transport, TransportDataHandler } from './Transport'
import { SerialPort } from 'serialport'
import { SlipDecoder, SlipEncoder } from '@serialport/parser-slip-encoder'

export class NodeSerialTransport implements Transport {
  private port: SerialPort | undefined
  private writer: SlipEncoder | undefined
  private dataHandler: TransportDataHandler | undefined
  private openHandler: (() => any) | undefined
  private closeHandler: (() => any) | undefined

  async open(options: any) {
    this.port = new SerialPort({
      baudRate: 9600,
      autoOpen: false,
      ...options,
    })

    this.port.on('close', () => this.closeHandler?.())

    this.setupReader()
    this.setupWriter()

    return new Promise<void>((resolve, reject) => {
      if (!this.port) throw new Error('Port is not initialized.')

      this.port.open((error) => {
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      })
    })
  }

  private setupReader() {
    if (!this.port) throw new Error('Port is not initialized.')

    const reader = this.port.pipe(new SlipDecoder())
    reader.on('data', (data: Buffer) => {
      this.dataHandler?.(new Uint8Array(data))
    })
  }

  private setupWriter() {
    this.writer = new SlipEncoder({
      // Send the END character also at the start (this is a special version
      // of SLIP that is implemented on the device).
      START: 0xc0,
    })
    this.port && this.writer.pipe(this.port)
  }

  write(data: Uint8Array) {
    return new Promise<void>((resolve, reject) => {
      this.writer?.write(Buffer.from(data), (error) => {
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      })
    })
  }

  onData(handler: TransportDataHandler) {
    this.dataHandler = handler
  }

  onOpen(handler: () => any) {
    this.openHandler = handler
  }

  onClose(handler: () => any) {
    this.closeHandler = handler
  }

  async close() {
    return new Promise<void>((resolve, reject) => {
      if (!this.port) throw new Error('Port is not initialized.')

      this.port.close((error) => {
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      })
    })
  }
}
