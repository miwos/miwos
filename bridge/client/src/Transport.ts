export interface Transport {
  write(data: Uint8Array): Promise<void>
  open(options?: any): Promise<void>
  close(): Promise<void>
  onData(handler: TransportDataHandler): any
  onOpen(handler: () => any): any
  onClose(handler: () => any): any
}

export type TransportDataHandler = (data: Uint8Array) => any