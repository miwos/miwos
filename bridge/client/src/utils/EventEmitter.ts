import { parsePathPattern, PathParams } from './parsePathPattern'

export type EventEmitterParams = PathParams

export type EventEmitterHandler = (
  payload: any,
  params: EventEmitterParams
) => any

export interface EventEmitterEvent {
  path: string
  match: (path: string) => PathParams | null
  handlers: Array<EventEmitterHandler>
}

/**
 * An event emitter that uses OSC style path's as event names.
 */
export class EventEmitter {
  events: { [path: string]: EventEmitterEvent } = {}

  on(path: string, handler: EventEmitterHandler) {
    const { events } = this
    events[path] = events[path] ?? {
      path,
      match: parsePathPattern(path),
      handlers: [],
    }
    events[path].handlers.push(handler)
  }

  off(path: string, handler: EventEmitterHandler) {
    const handlers = this.events[path]?.handlers
    const index = handlers?.indexOf(handler)
    if (index !== undefined && index >= 0) handlers.splice(index, 1)
  }

  once(path: string, handler: EventEmitterHandler) {
    const handlerOnce = (payload: any, params: PathParams) => {
      handler.call(this, payload, params)
      this.off(path, handlerOnce)
    }
    this.on(path, handlerOnce)
  }

  emit(path: string, payload?: any) {
    for (const { match, handlers } of Object.values(this.events)) {
      if (!handlers.length) continue
      const params = match(path)
      if (params) {
        handlers.forEach((handler) => handler.call(this, payload, params))
      }
    }
  }
}
