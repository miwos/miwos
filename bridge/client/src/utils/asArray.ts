// https://github.com/ditojs/dito/blob/master/packages/utils/src/base.js
export const asArray = (arg: any) =>
  Array.isArray(arg) ? arg : arg !== undefined ? [arg] : []
