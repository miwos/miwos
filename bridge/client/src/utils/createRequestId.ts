let requestId = -1

/**
 * Create a 16bit request id.
 */
export const createRequestId = () => {
  requestId++
  if (requestId > 65535) requestId = 0
  return requestId
}
