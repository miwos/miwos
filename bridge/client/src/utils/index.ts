export * from './asArray'
export * from './createRequestId'
export * from './EventEmitter'
export * from './parseDirList'
export * from './parsePathPattern'
