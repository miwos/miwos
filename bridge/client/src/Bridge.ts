import crc from 'crc/calculators/crc16xmodem'
// @ts-ignore (missing types)
import Message from 'osc-js'
import { Transport } from './Transport'
import { MessageArg } from './types'
import {
  asArray,
  createRequestId,
  EventEmitter,
  parseDirList,
  PathParams,
} from './utils'

export interface BridgeOptions {
  responseTimeout?: number
  debug?: boolean
}

export class Bridge extends EventEmitter {
  private readSerialMode: 'osc' | 'raw' = 'osc'
  private responseTimeout
  private debug

  constructor(
    private transport: Transport,
    { responseTimeout = 1000, debug = false }: BridgeOptions = {}
  ) {
    super()

    this.responseTimeout = responseTimeout
    this.debug = debug

    this.transport.onData((data: Uint8Array) => this.handleData(data))
    this.transport.onOpen(() => this.emit('/open'))
    this.transport.onClose(() => this.emit('/close'))

    this.on('/raw/**', () => (this.readSerialMode = 'raw'))

    if (this.debug) {
      this.on('/data/dev', (data) =>
        console.log('[debug] ' + new TextDecoder().decode(data))
      )
    }

    // Emit a normal log event for any raw log, so we have a unified API for
    // handling logs.
    this.on('/raw/log/:type', async (_, { type }) => {
      const data = await this.waitForRawData()
      const text = new TextDecoder().decode(data as Uint8Array)
      this.emit(`/log/${type}`, { args: [text] })
    })
  }

  open(options: any) {
    return this.transport.open(options)
  }

  close() {
    return this.transport.close()
  }

  notify(name: string, args: MessageArg | MessageArg[]) {
    this.sendMessage(name, args)
  }

  request(name: string, args: MessageArg | MessageArg[]) {
    const id = createRequestId()
    this.sendMessage(name, [id, ...asArray(args)])
    return this.waitForResponse(id)
  }

  async readFile(fileName: string) {
    const buffer = (await this.request('/file/read', fileName)) as Uint8Array
    const content = new TextDecoder().decode(buffer).slice('File:'.length)
    return content
  }

  async writeFile(fileName: string, content: string) {
    if (!content.length) throw new Error("file can't be empty")

    const id = createRequestId()
    const buffer = new TextEncoder().encode(content)
    const checkSum = crc(buffer)

    const parts = fileName.split('/')
    const baseName = parts.pop() ?? fileName
    const dirName = parts.join('/')

    await this.sendMessage('/raw/file/write', [id, dirName, baseName, checkSum])
    await this.transport.write(buffer)
    return this.waitForResponse(id)
  }

  removeFile(fileName: string) {
    return this.request('/file/remove', fileName)
  }

  removeDir(dirName: string) {
    return this.request('/dir/remove', dirName)
  }

  async getDir(dirName: string, recursive = false) {
    const buffer = await this.request('/dir/list', [dirName, recursive ? 1 : 0])
    const content = new TextDecoder().decode(buffer as any)

    // To prevent a timeout while waiting for an empty dir list, the device
    // always has to send something (a leading 'Dir:'). If we only received this
    // and nothing else the directory is empty.
    if (content === 'Dir:') return []

    return parseDirList(content.slice('Dir:'.length))
  }

  private sendMessage(name: string, args: MessageArg | MessageArg[]) {
    const message = new Message(name, ...asArray(args))
    const data = message.pack()
    return this.transport.write(data)
  }

  private waitForResponse(id: number): Promise<MessageArg> {
    return new Promise((resolve, reject) => {
      const timeout = this.startResponseTimeout(reject)

      // Listen to all responses until we find a response with the matching
      // request id.
      const handler = async (message: Message, params: PathParams) => {
        const [requestId, payload] = message.args

        if (requestId === id) {
          this.off('/r/:type', handler)
          this.off('/raw/r/:type', handler)
          clearTimeout(timeout)

          // If the response is raw, we only received the response's header
          // message and the actual raw serial data will follow.
          const isRaw = message.address.startsWith('/raw/')
          const response = isRaw ? await this.waitForRawData() : payload

          if (params.type === 'success') {
            resolve(response)
          } else {
            const message = isRaw
              ? new TextDecoder().decode(response)
              : response
            reject(new Error(message))
          }
        }
      }

      this.on('/r/:type', handler)
      this.on('/raw/r/:type', handler)
    })
  }

  private waitForRawData() {
    return new Promise((resolve, reject) => {
      const timeout = this.startResponseTimeout(reject)
      this.once('/data/raw', (payload: Uint8Array) => {
        clearTimeout(timeout)
        resolve(payload)
      })
    })
  }

  private startResponseTimeout(reject: (e: Error) => void) {
    return setTimeout(
      () => reject(new Error('Response timeout')),
      this.responseTimeout
    )
  }

  private handleData(data: Uint8Array) {
    // Might be useful during development.
    this.emit('/data/dev', data)

    if (this.readSerialMode === 'osc') {
      const message = new Message()
      try {
        message.unpack(new DataView(data.buffer))
      } catch {
        // The data wasn't a valid OSC message.
        this.emit('/data/unknown', data)
        return
      }
      this.emit(message.address, message)
    } else if (this.readSerialMode === 'raw') {
      this.emit('/data/raw', data)
      // All raw data comes as one chunk, so we don't expect any more raw data
      // an can switch back to osc mode.
      this.readSerialMode = 'osc'
    }
  }
}
