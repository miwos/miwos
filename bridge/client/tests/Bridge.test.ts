import { Bridge } from '../src'
import { NodeSerialTransport } from '../src/NodeSerialTransport'
import { dirIncludes, randomString } from './utils'

const testDir = '__test__'
const path = 'COM8'
const bridge = new Bridge(new NodeSerialTransport())

beforeAll(async () => {
  bridge.on('/log/:type', (message, { type }) => {
    const [text] = message.args
    ;(console as any)[type](text)
  })

  await bridge.open({ path })
  await bridge.removeDir(testDir)
})
afterAll(() => bridge.close())

describe('Bridge', () => {
  it('requests a response from the device', async () => {
    const response = await bridge.request('/echo/int', 123)
    expect(response).toBe(123)
  })

  it('handles request response timeouts', async () => {
    const nonExistingOscAddress = `/test/${randomString(50)}`
    await expect(bridge.request(nonExistingOscAddress, 0)).rejects.toThrow(
      'Response timeout'
    )
  })

  it('writes and reads a file', async () => {
    const content = performance.now().toString()
    const file = `${testDir}/test.txt`
    await bridge.writeFile(file, content)
    expect(await bridge.readFile(file)).toBe(content)
  })

  it('handles empty files', async () => {
    const file = `${testDir}/test.txt`
    await expect(bridge.writeFile(file, '')).rejects.toThrowError(
      "file can't be empty"
    )
  })

  it("it throws an error if a file doesn't exit", async () => {
    const nonExistingFile = `${testDir}/${Date.now()}.txt`
    await expect(bridge.readFile(nonExistingFile)).rejects.toThrowError(
      "file doesn't exist"
    )
  })

  it('it creates missing parent directories', async () => {
    const file = `${testDir}/deeply/nested/test.txt`
    await expect(bridge.writeFile(file, 'test')).resolves.not.toThrowError()
  })

  it('lists a directory', async () => {
    expect(await bridge.getDir(testDir)).toMatchSnapshot()
  })

  it('handles empty directories', async () => {
    const dir = `${testDir}/empty/`
    // Create a new empty directory by writing a file and then deleting it.
    await bridge.writeFile(`${dir}/test.txt`, 'test')
    await bridge.removeFile(`${dir}/test.txt`)
    expect(await bridge.getDir(dir)).toEqual([])
  })

  it('removes a file', async () => {
    const name = 'test.txt'
    const file = `${testDir}/${name}`
    await bridge.writeFile(file, 'test')
    await bridge.removeFile(file)

    const dir = await bridge.getDir(testDir)
    expect(dirIncludes(dir, name)).toBeFalsy()
  })

  it('removes a directory (recursively)', async () => {
    await bridge.removeDir(testDir)

    const root = await bridge.getDir('/')
    expect(dirIncludes(root, testDir)).toBeFalsy()
  })

  it('handles many consecutive file writes', async () => {
    const content = randomString(2000)
    for (let i = 0; i < 10; i++) {
      await bridge.writeFile(`${testDir}/many/file-${i}.txt`, content)
    }
    const dir = await bridge.getDir(`${testDir}/many`)
    expect(dir).toMatchSnapshot()
  })
})
