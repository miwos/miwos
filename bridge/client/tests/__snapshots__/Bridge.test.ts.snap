// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bridge handles many consecutive file writes 1`] = `
Array [
  Object {
    "name": "file-0.txt",
    "type": "file",
  },
  Object {
    "name": "file-1.txt",
    "type": "file",
  },
  Object {
    "name": "file-2.txt",
    "type": "file",
  },
  Object {
    "name": "file-3.txt",
    "type": "file",
  },
  Object {
    "name": "file-4.txt",
    "type": "file",
  },
  Object {
    "name": "file-5.txt",
    "type": "file",
  },
  Object {
    "name": "file-6.txt",
    "type": "file",
  },
  Object {
    "name": "file-7.txt",
    "type": "file",
  },
  Object {
    "name": "file-8.txt",
    "type": "file",
  },
  Object {
    "name": "file-9.txt",
    "type": "file",
  },
]
`;

exports[`Bridge lists a directory 1`] = `
Array [
  Object {
    "children": Array [],
    "name": "deeply",
    "type": "directory",
  },
  Object {
    "name": "test.txt",
    "type": "file",
  },
]
`;
